### ISSUE TITLE
How can I use titans in Continuous Time series?

### ISSUE THREAD

**Issue by hanbuck30** [lucidrains/titans-pytorch#27]
Thank you for your code.
I want to use titans  at EEG dataset.
How can I modify code to use EEG?

**Comment by nielth** [comment]
Did you find a solution for this @hanbuck30?

**Comment by hanbuck30** [comment]
I used EEG Dataset. I modified mac_transformer.py to include input and output projections and redefined the data shape to (batch, time, feature). Additionally, I adjusted the hyperparameters to better suit time series data. However, I am currently troubleshooting an error occurring in the Neural Memory component. @nielth

**Comment by nielth** [comment]
Would you let me know if you decide to commit your code to GitHub. I am doing a thesis on time series churn and it would be nice to see this being used in a practical environment. Thx. @hanbuck30

**Comment by lucidrains** [comment]
@hanbuck30 got it working? are you in your "don't scoop me" mode?

**Comment by hanbuck30** [comment]
I modified the MAC transformer's embedding layer, input shape, logits, and loss function. I'm not sure if these changes have been properly adapted for continuous data, but applying the hyperparameters provided for training MAC led to problems. It appears that the issue arose because I didn't thoroughly examine the neural memory.
@lucidrains

**Comment by lucidrains** [comment]
@hanbuck30 could you show me an example of how you input the continuous data into MAC and what you are trying to predict? i'll just build it for you

**Comment by hanbuck30** [comment]
The input data consists of EEG data, and the goal is to generate a Mel spectrogram from it. The EEG data is structured as (Batch, Time, Channels) and the mel spectrogram is also structured as (Batch, Time, Channels).

data = torch.rand(468, 500, 130)
mel = torch.rand(468, 172, 80)

def cycle(loader):
    while True:
        for data, target in loader:
            yield data.cuda()
            yield target.cuda()

train_dataset = TensorDataset(data[:400], mel[:400])
val_dataset = TensorDataset(data[400:], mel[400:])
train_loader = cycle(DataLoader(train_dataset, batch_size=32))
val_loader = cycle(DataLoader(val_dataset, batch_size=8))
# (The train and validation datasets/loaders are re-created here, which might be redundant)

# I added a Conv1D layer to adjust the time length.
NEURAL_MEMORY_DEPTH = 2
NUM_PERSIST_MEM = 4
NUM_LONGTERM_MEM = 4
NEURAL_MEM_LAYERS = (2, 4, 6)                   # Layers 2, 4, 6 include neural memory; you can add more.
NEURAL_MEM_GATE_ATTN_OUTPUT = False
NEURAL_MEM_MOMENTUM = True
NEURAL_MEM_MOMENTUM_ORDER = 1
NEURAL_MEM_QK_NORM = True
NEURAL_MEM_MAX_LR = 1e-1
USE_MEM_ATTENTION_MODEL = False
WINDOW_SIZE = 32
NEURAL_MEM_SEGMENT_LEN = 4                      # Set smaller for more granularity for learning rate/momentum, etc.
NEURAL_MEM_BATCH_SIZE = 128                     # Set smaller to update the neural memory weights more often as it traverses the sequence.
SLIDING_WINDOWS = True
STORE_ATTN_POOL_CHUNKS = True                   # Whether to use attention pooling for chunk-derived momentum, per-layer LR modulation, and decay.
MEMORY_MODEL_PER_LAYER_LEARNED_LR = True
NEURAL_MEM_WEIGHT_RESIDUAL = True               # Learning to incorporate contributions from the previous neural memory layer's weights yields significant improvements.
                                                # This was improvised and is not in the paper, but is inspired by the value residual learning free lunch paper.

NUM_BATCHES = int(1e5)
BATCH_SIZE = 4
GRADIENT_ACCUMULATE_EVERY = 4
LEARNING_RATE = 2e-4
VALIDATE_EVERY = 100
GENERATE_EVERY = 500
PRIME_LENGTH = 100
GENERATE_LENGTH = 512
SHOULD_GENERATE = True

USE_ACCELERATED_SCAN = True
USE_FLEX_ATTN = True
USE_FAST_INFERENCE = False

if USE_MEM_ATTENTION_MODEL:
    neural_memory_model = MemoryAttention(
        dim=64
    )
else:
    neural_memory_model = MemoryMLP(
        dim=64,
        depth=NEURAL_MEMORY_DEPTH
    )

mel = torch.rand(468, 172, 80)
# from titans_pytorch import MemoryAsContextTransformer
model = MemoryAsContextTransformer(
    input_dim=130,
    output_dim=80,
    kernel_size=3,
    stride=3,
    padding=9,
    dim=130,
    depth=8,
    segment_len=172, 
    num_persist_mem_tokens=NUM_PERSIST_MEM,
    num_longterm_mem_tokens=NUM_LONGTERM_MEM,
    neural_memory_layers=NEURAL_MEM_LAYERS,
    neural_memory_segment_len=NEURAL_MEM_SEGMENT_LEN,
    neural_memory_batch_size=NEURAL_MEM_BATCH_SIZE,
    neural_mem_gate_attn_output=NEURAL_MEM_GATE_ATTN_OUTPUT,
    neural_mem_weight_residual=NEURAL_MEM_WEIGHT_RESIDUAL,
    use_flex_attn=USE_FLEX_ATTN,
    sliding_window_attn=SLIDING_WINDOWS,  # local attention window size
    neural_memory_model=neural_memory_model,
    neural_memory_kwargs=dict(
        dim_head=64,
        heads=4,
        attn_pool_chunks=STORE_ATTN_POOL_CHUNKS,
        qk_rmsnorm=NEURAL_MEM_QK_NORM,
        momentum=NEURAL_MEM_MOMENTUM,
        momentum_order=NEURAL_MEM_MOMENTUM_ORDER,
        default_step_transform_max_lr=NEURAL_MEM_MAX_LR,
        use_accelerated_scan=USE_ACCELERATED_SCAN,
        per_parameter_lr_modulation=MEMORY_MODEL_PER_LAYER_LEARNED_LR
    )
).cuda()

# Optimizer
optim = AdoptAtan2(model.parameters(), lr=LEARNING_RATE)

loss = model(next(train_loader), target=mel, return_loss=True)

I used the provided code, but I'm not entirely sure how it should be utilized.

**Comment by lucidrains** [comment]
is the mel from the eeg? why not just have the input also in mel space? and is the objective autoregressive? (one time bin of the eeg predicting the mel of the same eeg or some other signal of the next)

**Comment by hanbuck30** [comment]
@lucidrains  No, the goal is to convert EEG input into the Mel-spectrogram of the ground truth voice. In the dataset, I used both EEG and Mel so that the model can learn to transform EEG into Mel, and I employed an L1 loss. Originally, the objective was to perform regression on the entire input at once, but now I'm considering an auto-regressive approach. It would be great to have a configurable time lag, and if possible, I'd like to experiment with both versions.

**Comment by lucidrains** [comment]
@hanbuck30 ah i see, yea, let me think about this being putting pen to paper

**Comment by lucidrains** [comment]
shouldn't the time dimension be divisible by the  mel time bins?

**Comment by hanbuck30** [comment]
@lucidrains I think so too!

**Comment by hanbuck30** [comment]
@lucidrains I'm considering a method to downsample EEG to match the Mel time bins using a Conv1D filter.

**Comment by hanbuck30** [comment]
@lucidrains Thank you for your feedback. While I have a general understanding of the theory behind mac_transformer and Neural Memory, I’ve found it challenging to grasp some parts of the internal code due to the various combined techniques. I’m currently modifying the project setup to apply it to an EEG dataset, and I was wondering if you could provide some advice on hyperparameter configuration when using Neural Memory.

Additionally, my professor mentioned that a one-shot regression approach tends to work better empirically than the auto-regressive method. Therefore, I plan to first explore using convolutional filters and, if necessary, consider interpolation as well. Thank you.
